Add pre-commit support, and fix end of line and file issues

Twon opened the pull request:
Adds support for pre-commit originally proposed in https://github.com/libfn/functional/pull/117, moved here due to the issues encountered with the merge against the rewritten history on main.

Bronek commented on README.md:
```diff
@@ -49,6 +49,24 @@ The maintainers are unable to guarantee that no significant refactoring will eve
 [ripple]: https://ripple.com/
 
+## Pre-commit
+This repository use [pre-commit](https://pre-commit.com/) to install git commit hooks which will run checks on the repository (currently just clang-format).  To install the git commit hooks locally so they run on each commit you make just set up a virtual environment, install the requirement and run pre-commit install:
```
Pls replace "use" with "supports", also list other check i.e. `pre-commit-hooks`.

The second sentence seem to be missing punctuation marks, but perhaps it's just me.

Twon replied:
Done, I've also reworded and broken up the second sentence.  Hopefully, that reads better now.

Bronek commented on README.md:
```diff
@@ -49,6 +49,24 @@ The maintainers are unable to guarantee that no significant refactoring will eve
+# Set up a virtual environment to install pre-commit
+python3 -m venv .venv
+source ./.venv/bin/activate
+pip install -r ./requirements.txt
```
I'd rather do without `requirements.txt` for the time being, just `pip install pre-commit`

Twon replied:
Done

Bronek commented on requirements.txt:
Let's not add this file - it may make readers confused (making this project look like it actually requires Python). If we were to add a workflow with an image to use `pre-commit` internally (as #117 suggested), then we would probably add this file under `ci/build` or similar, but at this time it's not needed.

Bronek reviewed (commented):
Thank you a lot for this change, also for cleaning up both line and file endings.

Several small comments; also sorry for making the `main` branch unstable today :crying_cat_face:

Twon commented:
> Thank you a lot for this change, also for cleaning up both line and file endings.
> 
> Several small comments; also sorry for making the `main` branch unstable today 😿

You're welcome.  I've addressed the PR comments.  Please let me know if there is anything else and I will address it.

Bronek commented:
T

> > Thank you a lot for this change, also for cleaning up both line and file endings.
> > Several small comments; also sorry for making the `main` branch unstable today 😿
> 
> You're welcome. I've addressed the PR comments. Please let me know if there is anything else and I will address it.

Thank you ! Before I merge it, I need commits to be signed. https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

Bronek reviewed (approved):
nice, thank you !
